src/simulation/dynamics/constraintEffector/_UnitTest/test_ConstraintDynamicEffectorUnit.py

```diff
@@ -0,0 +1,331 @@
+    combinedOrbAngMom = orbAngMom1_N + orbAngMom2_N
+    combinedOrbEnergy = orbEnergy1 + orbEnergy2
+    combinedRotAngMom = rotAngMom1PntCT_N + rotAngMom2PntCT_N
+    combinedRotEnergy = rotEnergy1 + rotEnergy2
+
+    # Plotting
+    plt.close("all")
+    plt.figure()
+    plt.clf()
+    for i in range(3):
+        plt.semilogy(constraintTimeData, np.abs(psi_B1[:, i]))
+    plt.semilogy(constraintTimeData, np.linalg.norm(psi_B1,axis=1))
+    plt.legend([r'$\psi_1$',r'$\psi_2$',r'$\psi_3$',r'$\psi$ magnitude'])
+    plt.xlabel('time (seconds)')
+    plt.ylabel(r'variation from fixed position: $\psi$ (meters)')
+    plt.title('Direction Constraint Violation Components')
+
+    plt.figure()
+    plt.clf()
+    for i in range(3):
+        plt.semilogy(constraintTimeData, np.abs(4*np.arctan(sigma_B2B1[:, i]) * macros.R2D))
+    plt.semilogy(constraintTimeData, np.linalg.norm(4*np.arctan(sigma_B2B1) * macros.R2D,axis=1))
+    plt.legend([r'$\phi_1$',r'$\phi_2$',r'$\phi_3$',r'$\phi$ magnitude'])
+    plt.xlabel('time (seconds)')
+    plt.ylabel(r'relative attitude angle: $\phi$ (deg)')
+    plt.title('Attitude Constraint Violation Components')
+
+    plt.figure()
+    plt.clf()
+    plt.plot(conservationTimeData, (combinedRotAngMom[:,0] - combinedRotAngMom[0,0])/combinedRotAngMom[0,0],
+             conservationTimeData, (combinedRotAngMom[:,1] - combinedRotAngMom[0,1])/combinedRotAngMom[0,1],
+             conservationTimeData, (combinedRotAngMom[:,2] - combinedRotAngMom[0,2])/combinedRotAngMom[0,2])
+    plt.xlabel('time (seconds)')
+    plt.ylabel('Relative Difference')
+    plt.title('Combined Rotational Angular Momentum')
+
+    plt.figure()
+    plt.clf()
+    plt.plot(conservationTimeData, (combinedRotEnergy - combinedRotEnergy[0])/combinedRotEnergy[0])
+    plt.xlabel('time (seconds)')
+    plt.ylabel('Relative Difference')
+    plt.title('Combined Rotational Energy')
+
+    if testCase == 'Gravity':
+        plt.figure()
+        plt.clf()
+        plt.plot(conservationTimeData, (combinedOrbAngMom[:,0] - combinedOrbAngMom[0,0])/combinedOrbAngMom[0,0], 
+             conservationTimeData, (combinedOrbAngMom[:,1] - combinedOrbAngMom[0,1])/combinedOrbAngMom[0,1], 
+             conservationTimeData, (combinedOrbAngMom[:,2] - combinedOrbAngMom[0,2])/combinedOrbAngMom[0,2])
+        plt.xlabel('time (seconds)')
+        plt.ylabel('Relative Difference')
+        plt.title('Combined Orbital Angular Momentum')
+
+        plt.figure()
+        plt.clf()
+        plt.plot(conservationTimeData, (combinedOrbEnergy - combinedOrbEnergy[0])/combinedOrbEnergy[0])
+        plt.xlabel('time (seconds)')
+        plt.ylabel('Relative Difference')
+        plt.title('Combined Orbital Energy')
+
+    if show_plots:
+        plt.show()
+    plt.close("all")
+
+    # Testing setup
+    dirCnstAccuracy = 1E-7
+    attCnstAccuracy = 1E-4
```

Comment:
Is this the best we can do?

Reply:
This is likely a function of the evolution speed of the unit test.  We don't have to show the limit of how well this math works, we just need to run a test to validate we are getting  expected results.

Reply:
The constraint violation accuracy tolerances are a function of several factors including the gains, how dynamic the spacecraft movement is, what external forces and torques there are, etc. So these accuracies are included to effectively monitor that the perforce isn't worsened by the introduction of future changes, not to verify some theoretical threshold. This is already pointed out in the PR description, but I've now added that justification to the unit test RST now to make it clear there.